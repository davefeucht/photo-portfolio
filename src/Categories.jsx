import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import Category from './Category.jsx';

class Categories extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      categories: [],
      showAllCategories: true,
      singleCategoryToShow: {id: 0, name: ""},
      errorMsg: ""
    }
  }

  _getCategories() {
    axios.get('http://' + this.props.site + '/wp-json/wp/v2/categories?exclude=1')
    .then(res => {
      const categories = res.data;
      this.setState({ categories });
    }, error => {
      const errorMsg = 'Did not work: ' + (error.response ? error.response : error);
      this.setState({ errorMsg });
    });
  }

  _showSpecificCategory(categoryId, categoryName) {
    this.setState({showAllCategories: !this.state.showAllCategories}); 
    this.setState({singleCategoryToShow: {id: categoryId, name: categoryName}}); 
  }

  componentWillMount() {
    this._getCategories();
  }
  
  render() {
    let categoryList = [];
    if(this.state.showAllCategories) {
      categoryList = this.state.categories.map(category =>
        { return ( <Category key={category.id.toString()} id={category.id} name={category.name} site={this.props.site} clickCategory={this._showSpecificCategory.bind(this)} showAllPosts={false} /> ); }
      );
    }
    else {
      categoryList = <Category key={this.state.singleCategoryToShow.id.toString()} id={this.state.singleCategoryToShow.id} name={this.state.singleCategoryToShow.name} site={this.props.site} clickCategory = {this._showSpecificCategory.bind(this)} showAllPosts={true} />;
    }
    return (
      <div className="category-list">
        {categoryList} 
        <div>{this.state.errorMsg}</div>
      </div>
    );
  }
}

export default Categories;
