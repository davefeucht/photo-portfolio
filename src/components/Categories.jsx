/*****************
* Categories component displays a list of existing photo categories 
* and makes them clickable to display just a single category.
*****************/

import React from "react";
import axios from "axios";
import CategoryList from "./styledComponents/CategoryList.jsx";
import CategoryThumbnail from "./CategoryThumbnail.jsx";
import PropTypes from "prop-types";

export default class Categories extends React.Component {

  static propTypes = {
    site: PropTypes.string,
    showSingleCategory: PropTypes.func
  };

  /* Add Default Props */

  //Set up state variables for list of categories, whether to show all categories, and if not, which
  //single category to show.
  //NOTE: Take list of categories out of state and store in a member variable.
  state = {
    categories: [],
    errorMsg: ""
  };

  _isMounted = false;

  //Method to get all existing categories
  _getCategories() {
    const getCategoriesRequest = `https://${this.props.site}/wp-json/wp/v2/categories?exclude=175`;
    if(this._isMounted) {
      axios.get(getCategoriesRequest)
        .then(res => {
          const categories = res.data;
          this.setState({ categories });
        })
        .catch(error => {
          const errorMsg = "Did not work: " + (error.response ? error.response : error);
          this.setState({ errorMsg });
        });
    }
  }

  _showSpecificCategory(categoryId, categoryName) {
    this.props.showSingleCategory(categoryId, categoryName);
  }

  //Just before the component is added to the DOM, get the list of categories.
  componentWillMount() {
    this._isMounted = true;
    this._getCategories();
  }

  componentWillUnmount() {
    this._isMounted = false;
  }
  
  render() {
    let categoryList = [];

    categoryList = this.state.categories.map(category =>
    { 
      //Display the Category component, and pass along the showSpecificCategory function as a prop, so that we can call it from the Category component
      return ( <CategoryThumbnail key={category.id.toString()} id={category.id} name={category.name} site={this.props.site} clickCategory={this._showSpecificCategory.bind(this)} /> ); }
    );

    return (
      <CategoryList>
        {categoryList} 
      </CategoryList>
    );
  }
}