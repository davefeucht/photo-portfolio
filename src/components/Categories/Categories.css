.category-list {
  min-height: 100%;
  display: grid;
  grid-template-columns: repeat(var(--number-of-columns), calc(100% / var(--number-of-columns)));
  grid-template-rows: repeat(var(--number-of-rows), 1fr);
}