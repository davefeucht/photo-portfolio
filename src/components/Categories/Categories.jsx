/*****************
* Categories component displays a list of existing photo categories 
* and makes them clickable to display just a single category.
*****************/

import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { reaction } from 'mobx';
import { observer } from 'mobx-react';
import CategoryThumbnail from '../CategoryThumbnail/CategoryThumbnail.jsx';
import './Categories.css';

const Categories = ({ stateStore, api }) => {
  const setRows = () => {
    const numberOfColumns = document.body.style.getPropertyValue('--number-of-columns');
    const numberOfCategories = stateStore.categoryList.length;
    document.body.style.setProperty('--number-of-rows', numberOfCategories / numberOfColumns);
  };

  useEffect(() => {
    setRows();
    const disposer = reaction(
      () => [stateStore.screenInfo.width, stateStore.screenInfo.height],
      () => setRows()
    );
    return () => {
      disposer();
    }
  });

  const _mapCategoryList = () => {
      return stateStore.categoryList.map((category, index) => { 
        //Display the Category component, and pass along the showSpecificCategory function as a prop, so that we can call it from the Category component
        return ( <CategoryThumbnail key={category.id.toString()} id={category.id} index={index} name={category.name} stateStore={stateStore} api={api} /> ); }
      );
  }
  
  return (
    <div className="category-list">
      {_mapCategoryList()} 
    </div>
  );
};

Categories.propTypes = {
  stateStore: PropTypes.object.isRequired,
  api: PropTypes.object.isRequired
};

Categories.displayName = 'Categories';

export default observer(Categories);