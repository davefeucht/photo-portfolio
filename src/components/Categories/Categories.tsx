/** ***************
* Categories component displays a list of existing photo categories 
* and makes them clickable to display just a single category.
**************** */

import './Categories.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect, useState } from 'react';

import { Category, ScreenInfo, SiteInfo } from '../../utils/types';
import CategoryThumbnail from '../CategoryThumbnail/CategoryThumbnail';
import PaginationNavigation from '../PaginationNavigation/PaginationNavigation';

interface CategoriesProps {
    maxItemsPerPage: number,
    screenInfo: ScreenInfo,
    categoryList: Category[]
}

const Categories: React.FC<CategoriesProps> = ({
    maxItemsPerPage,
    screenInfo,
    categoryList
}) => {
    const [currentPageIndex, setCurrentPageIndex] = useState<number>(1);

    const startIndex = (currentPageIndex - 1) * maxItemsPerPage;
    const endIndex = startIndex + maxItemsPerPage;

    const setRows = () => {
        const numberOfColumns = parseInt(document.body.style.getPropertyValue('--number-of-columns')) || 1;
        const numberOfCategories = maxItemsPerPage;
        document.body.style.setProperty('--number-of-rows', `${numberOfCategories / numberOfColumns}`);
    };

    useEffect(() => {
        setRows();
    }, [screenInfo.width, screenInfo.height]);

    return (
        <div className="category-list">
            {categoryList.slice(startIndex, endIndex)
                .map(category => {
                    return (<CategoryThumbnail key={category.id.toString()} id={category.id} name={category.name} />);
                })}
            <PaginationNavigation totalPages={categoryList.length / maxItemsPerPage} currentPageIndex={currentPageIndex} navigationFunction={(content: number) => { setCurrentPageIndex(content); }} />
        </div>
    );
};

Categories.displayName = 'Categories';

export default observer(Categories);
