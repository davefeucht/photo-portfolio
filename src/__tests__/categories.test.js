import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import Categories from '../components/Categories/Categories';
import stateStore from '../StateStore/store';

const categories = [
    {    
        id: 23,
        count: 6,
        description: 'Category of other things',
        link: '',
        name: 'Some other category',
        slug: 'smocg',
        taxonomy: '',
        parent: 12,
        meta: {}
    },
    {    
        id: 35,
        count: 3,
        description: 'Category of things',
        link: '',
        name: 'Some category',
        slug: 'smcg',
        taxonomy: '',
        parent: 23,
        meta: {}
    }
]

jest.mock('../utils/Api', () => ({
    getCategoryImage: () => Promise.resolve('https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg')
}));

test('Category list displays', async () => {
    const store = new stateStore();
    let component;
    let tree;
    store.setCategoryList(categories);
    await act(async () => {
        component = renderer.create(
            <MemoryRouter>
                <Categories stateStore={store}></Categories>
            </MemoryRouter>
        );

    })
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
