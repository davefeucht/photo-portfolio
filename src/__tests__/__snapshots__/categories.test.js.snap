// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category list displays 1`] = `
<div
  className="category-list"
>
  <a
    href="/category/23"
    onClick={[Function]}
  >
    <div
      className="category-thumbnail"
      style={
        {
          "backgroundImage": "url(https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg)",
        }
      }
    >
      <div
        className="category-title"
      >
        <div
          className="category-title-text"
        >
          Some other category
        </div>
      </div>
    </div>
  </a>
  <a
    href="/category/35"
    onClick={[Function]}
  >
    <div
      className="category-thumbnail"
      style={
        {
          "backgroundImage": "url(https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg)",
        }
      }
    >
      <div
        className="category-title"
      >
        <div
          className="category-title-text"
        >
          Some category
        </div>
      </div>
    </div>
  </a>
</div>
`;
